Minor fixes and small refactorings

camertron opened the pull request:
_**This pull request is part of a larger effort to bring OTF support to TTFunk. See https://github.com/prawnpdf/ttfunk/issues/53 for details.**_

Higlights of this PR:
1. Fix `Directory` so it parses the correct fields (I'm not even sure what `scaler_type` means or where it came from, I can't find it in any documentation).
2. Remove `POSTSCRIPT_GLYPH_MAPPING` constants from the Mac Roman and Windows 1252 subsets since they appear to not be used anywhere.
3. Take the head table's `index_to_loc_format` field into consideration when encoding.
4. Refactor the name table to be a little more object-oriented.

pointlessone commented on lib/ttfunk/table/name.rb:
```diff
@@ -88,55 +142,64 @@ def postscript_name
         count, string_offset = read(6, 'x2n*')
+        entries = []
 
-        @entries = []
```
There's a leftover `attr_reader` for this. Either remove that or keep this an ivar.

Related question: should it stay available for reading?

pointlessone commented on lib/ttfunk/table/name.rb:
```diff
@@ -88,55 +142,64 @@ def postscript_name
         end
 
-        @strings = Hash.new { |h, k| h[k] = [] }
+        @strings = []
```
This is a change in a public API.

pointlessone commented on lib/ttfunk/table/name.rb:
```diff
@@ -88,55 +142,64 @@ def postscript_name
+        @preferred_family = strings_by_name_id[16]
+        @preferred_subfamily = strings_by_name_id[17]
+        @compatible_full = strings_by_name_id[18]
+        @sample_text = strings_by_name_id[19]
```
I guess, all these numbers should get constants as well.

camertron commented:
> This is quite an extensive change to qualify as a "minor fixes and small refactorings".

Heh I suppose it is in isolation. It seems fairly insignificant compared to the rest of the changes.

> And it lacks any specs. I guess, you followed past development practices on TTFunk but those are not quite satisfactory when it comes to test coverage. I strongly encourage you to change it for better.

Right, I don't see much testing at all in TTFunk. Travis appears to only be running Rubocop, but maybe that's because I have some style violations which is causing `rake` to exit early. I will attempt to add some tests specifically for the `name` table and make sure existing tests pass.

> Along with passing test suite on TTFunk please also make sure at least Prawn and pdf-core gems are not broken by these changes.

Ok I can do that.

> On top of that please make sure you maintain code style.

Yes, absolutely.

camertron commented on .gitignore:
```diff
@@ -5,3 +5,5 @@
 Gemfile.lock
 doc
 *.gem
+.DS_Store
```
Ah good call.

camertron commented on lib/ttfunk/directory.rb:
```diff
@@ -1,11 +1,16 @@
+      # https://www.microsoft.com/typography/otspec/otff.htm#offsetTable
+      # We're ignoring searchRange, entrySelector, and rangeShift here, but
+      # skipping past them to get to the table information. Change the "Nn"
+      # to "Nn*" to decode those fields as well.
```
Do you have some sort of objection to leaving it in?

pointlessone replied:
No objection apart from the last sentence. It does not reflect current state of the code in any way.

camertron replied:
It provides additional knowledge to someone coming to this code later in case those fields are desired.

pointlessone replied:
I appreciate the care you show for fellow future developers but I'd rather have them read the spec and understand the reasoning behind the change they need to do than just follow instruction. As of now the last sentence of the comment doesn't provide any useful information about current state of the codebase.

Usually comments lag behind the code. This is the first time I see code lag behind the comment. 😄

camertron replied:
Lol ok, I'll take it out.

camertron commented on lib/ttfunk/table/cmap/format04.rb:
```diff
@@ -44,7 +44,6 @@ def self.encode(charmap)
           range_offsets = []
           glyph_indices = []
 
-          offset = 0
```
This may be the result of a bad merge. I will look into it.

camertron commented on lib/ttfunk/table/cmap/format04.rb:
```diff
@@ -56,15 +55,17 @@ def self.encode(charmap)
             end
 
-            start_glyph_id = new_map[a][:new]
+            start_glyph_id = new_map.fetch(a, new: 0)[:new]
```
Unfortunately I did this long enough ago that I don't fully remember the reasoning. If memory serves I was getting `NoMethodError`s, but those may have been fixed by [this](https://github.com/prawnpdf/ttfunk/pull/51) recently merged PR.

pointlessone replied:
It seems plausible. Could you please keep amount of change to the minimum required? It would make reviews a bit easier and a bit less code to maintain.

camertron commented on lib/ttfunk/table/head.rb:
```diff
@@ -21,10 +21,10 @@ class Head < TTFunk::Table
       attr_reader :glyph_data_format
 
-      def self.encode(head, loca)
+      def self.encode(head, loca = {})
```
This is to support the eventual addition of the `OtfEncoder` class, which won't be introduced for a while. TrueType files store glyphs in the `glyph` table and their locations are stored in the `loca` table, but OpenType files store glyphs as charstrings. The `OtfEncoder` class calls `Head.encode` with only one argument while `TtfEncoder` (also a future addition) calls it with a hash of `loca`s contents. By providing an optional second argument, both use cases can be supported. In the TrueType case, the `loca[:type]` value will be used; in the OpenType case, the original value from the `head` table will be used.

pointlessone replied:
Then maybe save this change for later when it's actually needed. It would make more sense that way in git history.

camertron replied:
Ok, I'll introduce it in one of the later branches.

camertron commented on lib/ttfunk/table/hmtx.rb:
```diff
@@ -22,7 +22,7 @@ def self.encode(hmtx, mapping)
 
       def for(glyph_id)
-        @metrics[glyph_id] ||
+        @metrics[glyph_id] ||=
```
Ahhh no it isn't, this is a misunderstanding on my part. Nice catch.

camertron commented:
Ok, I've gone through and fixed the Rubocop and test failures. I have also moved magic numbers into constants and tested against prawn. I ended up reverting the changes to the `name` table since string sorting (which is all I really wanted) has been fixed. I also changed `String` to `NameString` because seeing `class String < ::String` made me die a little inside.

I personally really dislike the `Layout/MultilineMethodCallIndentation` cop - is that something you and other maintainers would be willing to change? If not, then no worries :)

pointlessone commented:
> I personally really dislike the Layout/MultilineMethodCallIndentation cop - is that something you and other maintainers would be willing to change?

I'd like to keep Prawn style as close as possible to Ruby community code style (i.e. RuboCop default). For historical reasons current style a bit diverged. But this particular check seem to be inline with the community style.

Generally everything is up for debate as long as it's constructive but I'd like to keep style unchanged until after this whole thing lands to keep amount of unrelated changes to minimum.

pointlessone commented on lib/ttfunk/table/hmtx.rb:
```diff
@@ -22,15 +22,20 @@ def self.encode(hmtx, mapping)
-        @metrics[glyph_id] ||
-          HorizontalMetric.new(
+        metrics_cache[glyph_id] ||= begin
+          @metrics[glyph_id] || HorizontalMetric.new(
```
There's no need to cache the value that are readily available.

Try this:
```ruby
@metrics[glyph_id] || metrics_cache[glyph_id] ||= 
  HorizontalMetric.new(...)
```

camertron replied:
I'm trying to avoid creating `HorizontalMetric` instances when not necessary. In the cases where the metric is already available then yes, it will be cached by `metrics_cache`. In the case when it's not available, we create it and store it. I suppose we could do something like

```ruby
@metrics[glyph_id] || (metrics_cache[glyph_id] ||= HorizontalMetric.new(...))
```

Is that what you would prefer?

camertron replied:
Ah ok cool.

pointlessone commented on lib/ttfunk/table/name.rb:
```diff
@@ -45,10 +45,30 @@ def strip_extended
         tag = Digest::SHA1.hexdigest(key)[0, 6]
 
-        postscript_name = Name::String.new(
+        postscript_name = Name::NameString.new(
```
There's no need to namespace this any more, is there?

camertron replied:
Nope :)

camertron commented on Gemfile:
```diff
@@ -1,3 +1,5 @@
 
 gemspec
+
+gem 'pry-byebug', platform: :mri
```
This is only a dependency for someone who bundles inside the ttfunk repo itself, but I take your point and will remove it.

camertron commented:
@pointlessone What would you think about simply squash merging?

pointlessone commented:
That would work. GH makes ugly commit messages though. So if you don't mind could you please do it in the branch first?

camertron commented:
@pointlessone done :)
